docs/content/schemas.md

```diff
@@ -108,19 +127,215 @@ Consider the following input document:
+
+```
+package policy
+
+import data.acl
+
+default allow = false
+
+# METADATA
+# scope: rule
+# schemas:
+#   - data.acl: schemas.acl-schema
+allow {
+        access = data.acl["alice"]
+        access[_] == input.operation
+}
+
+allow {
+        access = data.acl["bob"]
+        access[_] == input.operation
+}
+```
+
+Consider a directory named `schemas` with the following structure, provided via `opa eval --schema opa-schema-examples/schemas`
+
+```$ tree schemas/
+schemas/
+├── default-input-schema.json
+└── acl-schema.json
+```
+
+For actual code samples, see: https://github.com/aavarghese/opa-schema-examples/acl
+
+In the first `allow` rule above, `data.acl` has the schema `schemas.acl-schema`. Note that we use the relative path inside the `schemas` directory to identify a schema and omit the `.json` suffix. 
+
+If we had the expression `data.acl.foo` in this rule, it would result in a type error because it is not allowed according to `schemas.acl-schema`.
+
+On the other hand, this annotation does not constrain other paths under `data`. What it says is that we know the type of `data.acl` statically, but not that of other paths. So for example, `data.foo` is not a type error and gets assigned the type `Any`.
+
+Additionally, the schema annotations specified in the Metadata for the first `allow` rule is also intentionally applied to the second `allow` rule (even though it doesn't have a METADATA comment block attached to it) since both rules are defined with the same name.
+
+### Using the default-input-schema in an annotation
+
+When passing a directory of schemas that contains `default-input-schema.json`, we associate this schema with the `input` document globally, meaning for all rules in all packages. It is possible to reuse this schema in an annotation to associate it with another expression. 
+This is done as follows:
+```
+# METADATA
+# scope: rule
+# schemas: 
+#   - <expression>: default-input-schema
+```
+
+Notice that in this case, we omit the relative path to the `default-input-schema.json` file, because it is a special case. This could be used for example to associate the same schema with a data document as well:
+```
+# METADATA
+# scope: rule
+# schemas: 
+#   - data.doc: default-input-schema
+```
+
+### Schema overriding
+
+JSON Schemas are often incomplete specifications of the format of data. For example, a Kubernetes Admission Review resource has a field `object` which can contain any other Kubernetes resource. A schema for Admission Review has a generic type `object` for that field that has no further specification. To allow more precise type checking in such cases, we support overriding existing schemas. 
+
+Consider the following example:
+```
+package kubernetes.admission                                                
+
+# METADATA
+# scope: rule
+# schemas: 
+#   - input: schemas.kubernetes.admission-review
```

Comment:
@tsandall we are going to adopt the convention that the user has to write these as a slash-separated file path.

So the example would be:
```
# METADATA
# scope: rule
# schemas: 
#   - input: schemas/kubernetes/admission-review
```

Reply:
Is there a reason not to use the reference syntax here? In my head, I was thinking `schemas` as a new top-level document like `data` and `input` and so it made sense to refer to it via the reference syntax but I'm open to either.

Reply:
So we came across this example recently which would be harder to write in the reference I believe:

```
# METADATA
# scope: rule
# schemas:
#   - data.lib.konstraint.core.resource: release-4.5-local.pod-v1
```

The top-level directory is ` release-4.5-local` and the schema there is `pod-v1`. In the slash-separated file name format, it would be easy and unambiguous:

```
# METADATA
# scope: rule
# schemas:
#   - data.lib.konstraint.core.resource: release-4.5-local/pod-v1
```

How would you write that as a reference?

Reply:
I was assuming that all schemas were going to be referenced under a `schema` global variable so this would be `schemas["release-4.5-local"]["pod-v1"]`.

Reply:
I think a global variable (let's call it `schema` singular, so we don't have to write the extra `s` every time 😅 ) would be good. It gives us a well-known root for all references to schema just like we have `input` and `data` today.

For your example:

```
opa eval -s X/Y/Z/release-4.5-local ...
```

Then the path to the `pod-v1` schema would look like this:

```
# METADATA
# scope: rule
# schemas:
# - data.lib.constraint.core.resource: schema["pod-v1"]
```

If we're loading schema directories recursively, we'll need to mount the files under `schema` accordingly. For example, if `pod-v1` came from `X/Y/Z/release-4.5-local/A/B/pod-v1.json` then the schema ref would be:

``` 
data.lib.constraint.core.resource: schema.A.B["pod-v1"]
```

This feels consistent w/ `data` loading except that the filename has meaning, however, since we're planning to source these schema files from various places, that seems like the right approach.

Reply:
This new syntax is implemented in the latest commits in this PR. Docs and examples in https://github.com/aavarghese/opa-schema-examples/tree/ruleAnnotation have been updated too. Thanks Torin.